File: test/game/TypingGame.cs
/*
 * 控制台打字游戏 - C#版本
 * 运行命令: dotnet run (需要先创建项目，见README.md)
 * 
 * C#是编译型语言，运行在.NET虚拟机上（类似Java的JVM）
 * 本程序展示了C#的现代语法：属性、LINQ、var类型推断、Lambda表达式
 */

// 【using指令】类似Python的import，导入命名空间
// 这些是.NET框架的核心命名空间
using System;                    // 基础类型（Console, String, Random等）
using System.Collections.Generic; // 泛型集合（List<T>, Dictionary<K,V>）
using System.Linq;               // LINQ查询语言（Where, Select等）
using System.Threading;          // 多线程（本例未使用但已导入）

// 【命名空间】组织代码的容器，类似Java的package
// 防止类名冲突，便于代码管理
namespace TypingGame
{
    class Program
    {
        static void Main(string[] args)
        {
            var game = new TypingGame();
            game.Start();
        }
    }

    public class TypingGame
    {
        private static readonly string[] Words = {
            "namespace", "using", "var", "dynamic", "async", "await", "get", "set",
            "public", "private", "static", "void", "class", "interface", "string",
            "int", "double", "bool", "List", "Dictionary", "Array", "Console",
            "Task", "Action", "Func", "Event", "Delegate", "LINQ", "Lambda",
            "print", "input", "def", "return", "if", "else", "for", "while",
            "try", "catch", "finally", "throw", "new", "this", "base", "null",
            "variable", "function", "method", "parameter", "argument",
            "array", "loop", "condition", "recursion", "algorithm",
            "debug", "compile", "runtime", "syntax", "exception",
            "git", "commit", "push", "pull", "merge", "branch",
            "api", "json", "database", "server", "frontend", "backend", "fullstack"
        };

        private readonly Random _random;
        private readonly GameStats _stats;

        public TypingGame()
        {
            _random = new Random();
            _stats = new GameStats();
        }

        public void Start()
        {
            while (true)
            {
                ClearScreen();
                PrintHeader();

                var settings = SelectDifficulty();
                PlayGame(settings);
                ShowResults();

                Console.Write("\n再玩一次? (y/n): ");
                var choice = Console.ReadLine()?.Trim().ToLower();
                if (choice != "y")
                {
                    Console.WriteLine("\n感谢游玩! 继续搬砖...");
                    break;
                }
            }
        }

        private void ClearScreen()
        {
            for (int i = 0; i < 50; i++)
                Console.WriteLine();
        }

        private void PrintHeader()
        {
            Console.WriteLine("==================================================");
            Console.WriteLine("  控制台打字游戏 - C#版 (.NET现代版)");
            Console.WriteLine("==================================================");
            Console.WriteLine();
        }

        private DifficultySettings SelectDifficulty()
        {
            Console.WriteLine("选择难度:");
            Console.WriteLine("  1. 简单 (60秒, 短词汇)");
            Console.WriteLine("  2. 中等 (60秒, 全部词汇)");
            Console.WriteLine("  3. 困难 (45秒, 长词汇)");
            Console.WriteLine();

            while (true)
            {
                Console.Write("输入数字 (1-3): ");
                var choice = Console.ReadLine()?.Trim();

                switch (choice)
                {
                    case "1":
                        return new DifficultySettings(60, GetShortWords());
                    case "2":
                        return new DifficultySettings(60, Words);
                    case "3":
                        return new DifficultySettings(45, GetLongWords());
                    default:
                        Console.WriteLine("无效输入，请重新选择");
                        break;
                }
            }
        }

        private string[] GetShortWords() => 
            Words.Where(w => w.Length <= 5).ToArray();

        private string[] GetLongWords() => 
            Words.Where(w => w.Length >= 6).ToArray();

        private void PlayGame(DifficultySettings settings)
        {
            ClearScreen();
            PrintHeader();

            Console.WriteLine($"时间限制: {settings.TimeLimit}秒");
            Console.WriteLine($"词汇数量: {settings.WordPool.Length}");
            Console.WriteLine("规则: 输入显示的单词，按Enter确认");
            Console.WriteLine("准备开始...");
            Console.Write("\n按Enter键开始!");
            Console.ReadLine();

            ClearScreen();
            var startTime = DateTime.Now;
            _stats.Reset();

            while (true)
            {
                var elapsed = DateTime.Now - startTime;
                if (elapsed.TotalSeconds >= settings.TimeLimit)
                    break;

                var word = settings.GetRandomWord(_random);
                var timeLeft = (int)(settings.TimeLimit - elapsed.TotalSeconds);

                Console.WriteLine($"\n[剩余时间: {timeLeft}秒 | 正确: {_stats.CorrectCount} | 已输入: {_stats.TotalCount}]");
                Console.WriteLine($"单词: {word}");
                Console.Write("> ");

                var userInput = Console.ReadLine()?.Trim();

                if (string.IsNullOrEmpty(userInput))
                    continue;

                _stats.RecordAttempt(word, userInput);

                if (userInput.Equals(word, StringComparison.OrdinalIgnoreCase))
                {
                    Console.WriteLine("✓ 正确!");
                }
                else
                {
                    Console.WriteLine($"✗ 错误! 正确答案是: {word}");
                }
            }
        }

        private void ShowResults()
        {
            ClearScreen();
            PrintHeader();

            Console.WriteLine("游戏结束!");
            Console.WriteLine("==================================================");
            Console.WriteLine($"正确: {_stats.CorrectCount} / {_stats.TotalCount}");
            Console.WriteLine($"准确率: {_stats.Accuracy:P1}");

            var wrongWords = _stats.GetWrongWords();
            if (wrongWords.Any())
            {
                Console.WriteLine("\n错误记录:");
                var displayCount = Math.Min(wrongWords.Count, 10);
                for (int i = 0; i < displayCount; i++)
                {
                    Console.WriteLine($"  {i + 1}. {wrongWords[i]}");
                }
                if (wrongWords.Count > 10)
                {
                    Console.WriteLine($"  ... 还有 {wrongWords.Count - 10} 个错误");
                }
            }

            Console.WriteLine();
            if (_stats.Accuracy >= 0.95 && _stats.CorrectCount >= 20)
            {
                Console.WriteLine("🏆 评级: 代码大神!");
            }
            else if (_stats.Accuracy >= 0.8 && _stats.CorrectCount >= 15)
            {
                Console.WriteLine("⭐ 评级: 优秀程序员!");
            }
            else if (_stats.Accuracy >= 0.6)
            {
                Console.WriteLine("👍 评级: 还在进步!");
            }
            else
            {
                Console.WriteLine("💪 评级: 继续练习!");
            }
            Console.WriteLine();
        }
    }

    public class DifficultySettings
    {
        public int TimeLimit { get; }
        public string[] WordPool { get; }

        public DifficultySettings(int timeLimit, string[] wordPool)
        {
            TimeLimit = timeLimit;
            WordPool = wordPool;
        }

        public string GetRandomWord(Random random) => 
            WordPool[random.Next(WordPool.Length)];
    }

    public class GameStats
    {
        public int CorrectCount { get; private set; }
        public int TotalCount { get; private set; }
        public double Accuracy => TotalCount > 0 ? (double)CorrectCount / TotalCount : 0;
        
        private readonly List<string> _wrongWords;

        public GameStats()
        {
            _wrongWords = new List<string>();
        }

        public void Reset()
        {
            CorrectCount = 0;
            TotalCount = 0;
            _wrongWords.Clear();
        }

        public void RecordAttempt(string correctWord, string userInput)
        {
            TotalCount++;
            if (userInput.Equals(correctWord, StringComparison.OrdinalIgnoreCase))
            {
                CorrectCount++;
            }
            else
            {
                _wrongWords.Add($"'{correctWord}' → 你输入: '{userInput}'");
            }
        }

        public IReadOnlyList<string> GetWrongWords() => _wrongWords;
    }
}

File: test/game/TypingGame.java
import java.util.Scanner;
import java.util.List;
import java.util.ArrayList;
import java.util.Random;

public class TypingGame {
    
    private static final String[] WORDS = {
        "public", "private", "static", "void", "class", "extends", "implements",
        "interface", "package", "import", "new", "this", "super", "null", "true", "false",
        "String", "Integer", "ArrayList", "HashMap", "System", "println", "Scanner",
        "print", "input", "def", "return", "if", "else", "elif", "for", "while",
        "try", "catch", "finally", "throw", "throws", "abstract", "final",
        "variable", "function", "method", "parameter", "argument", "return",
        "array", "loop", "condition", "recursion", "algorithm",
        "debug", "compile", "runtime", "syntax", "exception",
        "git", "commit", "push", "pull", "merge", "branch",
        "api", "json", "database", "server", "frontend", "backend"
    };
    
    private Scanner scanner;
    private Random random;
    private GameStats stats;
    
    public TypingGame() {
        this.scanner = new Scanner(System.in);
        this.random = new Random();
        this.stats = new GameStats();
    }
    
    public void start() {
        while (true) {
            clearScreen();
            printHeader();
            
            DifficultySettings settings = selectDifficulty();
            playGame(settings);
            showResults();
            
            System.out.print("\n再玩一次? (y/n): ");
            String choice = scanner.nextLine().trim().toLowerCase();
            if (!choice.equals("y")) {
                System.out.println("\n感谢游玩! 继续搬砖...");
                break;
            }
        }
    }
    
    private void clearScreen() {
        for (int i = 0; i < 50; i++) {
            System.out.println();
        }
    }
    
    private void printHeader() {
        System.out.println("==================================================");
        System.out.println("  控制台打字游戏 - Java版 (面向对象版)");
        System.out.println("==================================================");
        System.out.println();
    }
    
    private DifficultySettings selectDifficulty() {
        System.out.println("选择难度:");
        System.out.println("  1. 简单 (60秒, 短词汇)");
        System.out.println("  2. 中等 (60秒, 全部词汇)");
        System.out.println("  3. 困难 (45秒, 长词汇)");
        System.out.println();
        
        while (true) {
            System.out.print("输入数字 (1-3): ");
            String choice = scanner.nextLine().trim();
            
            switch (choice) {
                case "1":
                    return new DifficultySettings(60, getShortWords());
                case "2":
                    return new DifficultySettings(60, WORDS);
                case "3":
                    return new DifficultySettings(45, getLongWords());
                default:
                    System.out.println("无效输入，请重新选择");
            }
        }
    }
    
    private String[] getShortWords() {
        List<String> shortWords = new ArrayList<>();
        for (String word : WORDS) {
            if (word.length() <= 5) {
                shortWords.add(word);
            }
        }
        return shortWords.toArray(new String[0]);
    }
    
    private String[] getLongWords() {
        List<String> longWords = new ArrayList<>();
        for (String word : WORDS) {
            if (word.length() >= 6) {
                longWords.add(word);
            }
        }
        return longWords.toArray(new String[0]);
    }
    
    private void playGame(DifficultySettings settings) {
        clearScreen();
        printHeader();
        
        System.out.println("时间限制: " + settings.getTimeLimit() + "秒");
        System.out.println("词汇数量: " + settings.getWordPool().length);
        System.out.println("规则: 输入显示的单词，按Enter确认");
        System.out.println("准备开始...");
        System.out.print("\n按Enter键开始!");
        scanner.nextLine();
        
        clearScreen();
        long startTime = System.currentTimeMillis();
        stats.reset();
        
        while (System.currentTimeMillis() - startTime < settings.getTimeLimit() * 1000) {
            String word = settings.getRandomWord(random);
            long timeLeft = settings.getTimeLimit() - 
                (System.currentTimeMillis() - startTime) / 1000;
            
            System.out.println("\n[剩余时间: " + timeLeft + "秒 | 正确: " + 
                stats.getCorrectCount() + " | 已输入: " + stats.getTotalCount() + "]");
            System.out.println("单词: " + word);
            System.out.print("> ");
            
            String userInput = scanner.nextLine().trim();
            
            if (userInput.isEmpty()) {
                continue;
            }
            
            stats.recordAttempt(word, userInput);
            
            if (userInput.equalsIgnoreCase(word)) {
                System.out.println("✓ 正确!");
            } else {
                System.out.println("✗ 错误! 正确答案是: " + word);
            }
        }
    }
    
    private void showResults() {
        clearScreen();
        printHeader();
        
        System.out.println("游戏结束!");
        System.out.println("==================================================");
        System.out.println("正确: " + stats.getCorrectCount() + " / " + stats.getTotalCount());
        System.out.printf("准确率: %.1f%%\n", stats.getAccuracy() * 100);
        
        List<String> wrongWords = stats.getWrongWords();
        if (!wrongWords.isEmpty()) {
            System.out.println("\n错误记录:");
            int displayCount = Math.min(wrongWords.size(), 10);
            for (int i = 0; i < displayCount; i++) {
                System.out.println("  " + (i + 1) + ". " + wrongWords.get(i));
            }
            if (wrongWords.size() > 10) {
                System.out.println("  ... 还有 " + (wrongWords.size() - 10) + " 个错误");
            }
        }
        
        System.out.println();
        if (stats.getAccuracy() >= 0.95 && stats.getCorrectCount() >= 20) {
            System.out.println("🏆 评级: 代码大神!");
        } else if (stats.getAccuracy() >= 0.8 && stats.getCorrectCount() >= 15) {
            System.out.println("⭐ 评级: 优秀程序员!");
        } else if (stats.getAccuracy() >= 0.6) {
            System.out.println("👍 评级: 还在进步!");
        } else {
            System.out.println("💪 评级: 继续练习!");
        }
        System.out.println();
    }
    
    public static void main(String[] args) {
        TypingGame game = new TypingGame();
        game.start();
    }
}

class DifficultySettings {
    private int timeLimit;
    private String[] wordPool;
    
    public DifficultySettings(int timeLimit, String[] wordPool) {
        this.timeLimit = timeLimit;
        this.wordPool = wordPool;
    }
    
    public int getTimeLimit() {
        return timeLimit;
    }
    
    public String[] getWordPool() {
        return wordPool;
    }
    
    public String getRandomWord(Random random) {
        return wordPool[random.nextInt(wordPool.length)];
    }
}

class GameStats {
    private int correctCount;
    private int totalCount;
    private List<String> wrongWords;
    
    public GameStats() {
        this.wrongWords = new ArrayList<>();
    }
    
    public void reset() {
        correctCount = 0;
        totalCount = 0;
        wrongWords.clear();
    }
    
    public void recordAttempt(String correctWord, String userInput) {
        totalCount++;
        if (userInput.equalsIgnoreCase(correctWord)) {
            correctCount++;
        } else {
            wrongWords.add("'" + correctWord + "' → 你输入: '" + userInput + "'");
        }
    }
    
    public int getCorrectCount() {
        return correctCount;
    }
    
    public int getTotalCount() {
        return totalCount;
    }
    
    public double getAccuracy() {
        return totalCount > 0 ? (double) correctCount / totalCount : 0.0;
    }
    
    public List<String> getWrongWords() {
        return wrongWords;
    }
}
